### DISCUSSION TITLE
can use rest api to create a dag or another way?

### DISCUSSION THREAD

**Discussion by tonglin96** [apache/airflow#20413] [Q&A] [answered]
can use rest api to create a dag or another way?

**Comment by manuzhang** [comment, marked as answer]
You may try https://github.com/teamclairvoyant/airflow-rest-api-plugin

**Comment by potiuk** [comment]
However the only "really" good way is to generate a DAG file so that scheduler can parse it.

**Reply by manuzhang** [reply]
That's what the airflow-rest-api-plugin does.

**Reply by potiuk** [reply]
Surely. You can rely on a 1200+ lines of code plugin for which code was last updated 2 years ago and might or might not work with the current version of airlfow (likely not).. 

My point is however that it's better to rely on the way that is guaranteed to work. Depending on your filesytem it's usually as easy as the existing filesystem functionality of copying a a file (depending on your filesystem deployment). Usually it's one line of code.

**Reply by manuzhang** [reply]
That's why I said "You may try" earlier. Meanwhile, We are using a modified version that's working with 2.1.4 and our usage is to create dags from another service based on user inputs.

**Reply by tonglin96** [reply]
> That's why I said "You may try" earlier. Meanwhile, We are using a modified version that's working with 2.1.4 and our usage is to create dags from another service based on user inputs.

Thank you very much for the repo you provided for my reference. my needs are also create dags from another service based on user inputs

**Reply by manuzhang** [reply]
@potiuk It will be great if Airflow can natively support uploading a DAG file through REST API.

**Reply by potiuk** [reply]
It will. This is not currently possible due to the architecture we have. 
But we are working on Airflow Improvement proposals that will make it possible. After the two AIP-s:
* https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-43+DAG+Processor+separation
* https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-44+Airflow+Internal+API

Another prerequisite is to serialize the full "executable" DAG to the database - whic is something AIrBnb is  going to write an AIP about soon most likely.

We will have an internal API server that will be able to process externally provided DAGs and one of the follow-up AIPs (not yet planned) will likely open it up for externally provided DAGs. But that's quite a bit later phase. Don't count on it too quickly.

**Reply by potiuk** [reply]
> That's why I said "You may try" earlier. Meanwhile, We are using a modified version that's working with 2.1.4 and our usage is to create dags from another service based on user inputs.

Even that is as easy now as saving prepared DAG file to DAG folder. Lots of people are doing it this way. No need to go through API.

There is really no need to use REST API for that. An API will really become useful if /when we change the model of Airlfow where DAGs are not scanned from the folder, but when you will be able to submit a DAG and simply save it to the DB and execute it without having to provide shared folders between components of Airlfow. Then API will be really useful. Before that - let me repeat it again - adding a REST API component that does not do anything else than writing file to the shared folder is kinda useless. It is far easier to open up the shared folder (or git sync or whatever) to write externally to rather than going through Airflow Webserver. Storing a file in shared folder is very simple comparing to having to got through an API.  It's just using the usual file API which is super-simple.

**Reply by manuzhang** [reply]
Thanks for sharing the plan and proposals.

> An API will really become useful if /when we change the model of Airflow where DAGs are not scanned from the folder, but when you will be able to submit a DAG and simply save it to the DB and execute it without having to provide shared folders between components of Airlfow. 

I agree. Arguably, airflow-rest-api-plugin provides an illusion of that without introducing other dependencies into my service. Anyway, I'm looking forward to the new API, and enhanced APIs to query DAGs with filters, update and delete.

**Reply by potiuk** [reply]
I like the word "illusion" :) . It is very much this. I see your point about dependencies. It is a question of perspective and limitations you have.

However, I am quite sure that injecting 1200+ lines of code in a big project like Airflow is much more brittle than simply using a tool/library to sync files (dedicated for that). 

You have no idea on performance impact of this plugin for Webserver. Maybe Airlfow webserver will get  10x  slower in some cases with it ? Maybe it opens up a security hole ? Maybe it breaks other functionality of the webserver? Maybe it blocks other users from interacting in airflow UI while you are sending the files ? You never know really. It sounds really scary actually.
